compiler/dialects/accfg.md

```diff
@@ -0,0 +1,141 @@
+# `accfg` Dialect Documentation
+
+The `accfg` dialect is meant to server as a genric optimisation dialect, which may be used by any lowering pipieline
+who has certain semantics and wants to use our optimisations. It is therefore paramount that the dialect documents
+the semantics of its operations and the assumptions made by its passes.
+
+## Operation Semantics
+
+The `accfg` dialect provides three operations that interact with a stateful "client" device. These operations are as
+follows:
+
+- `accfg.setup` writes to zero or more configuration registers of the client device
+- `accfg.launch` signals the client to start an execution that may be asynchronous
+- `accfg.await` blocks until the client device is finished executing
+
+Each operation identifies the client it's interacting with uniquely through a compile-time known property called
+`accelerator`. **The client must persist values in its configuration registers across `launch` invocations.**
+It is also assumed that all sources that effect the accelerators registers is expressed in the `accfg` dialect.
+
+Every `setup` operation produces an `accfg.state` SSA value, representing the current state of the accelerators
+registers after the setup operation acted on it. The parameters set by this operation are called *setup parameters*.
+A setup may take another `state` value as input. If provided, the output state represents the input state updated
+with the setup parameters of this operation. A setup acts destructively on the input state, meaning the input state
+**must not** be used after it has been consumed by another `setup` op. If no input state is provided, no assumptions
+can be made about the state before the setup operation.
+
+The `launch` operation takes a `state` as input, representing the configuration that it expects to launch. The launch
+can be provided with additional configuration parameters, which we call *launch parameters*. The `launch` will produce a
+`token` SSA value, which must be consumed by a corresponding `await` operation.
+
+
+To summarise, our semantics are:
+- Only one `accfg.state` variable per client can be alive at any given time.
+- Nothing but `accfg` dialect operations modify the accelerator state.
```

Comment:
This seems to contradict with #166 ?

Reply:
Hmm, yes. This phrasing should be improved.